embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/startup/AppStartupTraceEmitter.kt

```diff
@@ -0,0 +1,304 @@
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Records startup traces based on the data that it is provided. It adjusts what it logs based on what data has been provided and
+ * and which Android version the app is running.
+ */
+internal class AppStartupTraceEmitter(
+    private val clock: Clock,
+    private val startupServiceProvider: Provider<StartupService?>,
+    private val spanService: SpanService,
+    private val backgroundWorker: BackgroundWorker,
+    private val versionChecker: VersionChecker,
+    private val logger: InternalEmbraceLogger
+) {
+    private val processCreateRequestedMs: Long?
+    private val processCreatedMs: Long?
+
+    init {
+        val timestampAtDeviceStart = nowMs() - clock.nanoTime().nanosToMillis()
+        processCreateRequestedMs = if (versionChecker.isAtLeast(VERSION_CODES.N)) {
+            timestampAtDeviceStart + Process.getStartElapsedRealtime()
+        } else {
+            null
+        }
+        processCreatedMs = if (versionChecker.isAtLeast(VERSION_CODES.TIRAMISU)) {
+            timestampAtDeviceStart + Process.getStartRequestedElapsedRealtime()
+        } else {
+            null
+        }
+    }
+
+    @Volatile
+    private var applicationInitStartMs: Long? = null
+
+    @Volatile
+    private var applicationInitEndMs: Long? = null
+
+    @Volatile
+    private var startupActivityInitStartMs: Long? = null
+
+    @Volatile
+    private var startupActivityInitEndMs: Long? = null
+
+    @Volatile
+    private var startupActivityResumedMs: Long? = null
+
+    @Volatile
+    private var firstFrameRenderedMs: Long? = null
+
+    private val startupRecorded = AtomicBoolean(false)
+    private val endWithFrameDraw: Boolean = versionChecker.isAtLeast(VERSION_CODES.Q)
+
+    fun applicationInitStart(timestampMs: Long? = null) {
+        applicationInitStartMs = timestampMs ?: nowMs()
+    }
+
+    fun applicationInitEnd(timestampMs: Long? = null) {
+        applicationInitEndMs = timestampMs ?: nowMs()
+    }
+
+    fun startupActivityInitStart(timestampMs: Long? = null) {
+        if (startupActivityInitStartMs == null) {
+            startupActivityInitStartMs = timestampMs ?: nowMs()
+        }
+    }
+
+    fun startupActivityInitEnd(timestampMs: Long? = null) {
+        startupActivityInitEndMs = timestampMs ?: nowMs()
+    }
+
+    fun startupActivityResumed(timestampMs: Long? = null) {
+        startupActivityResumedMs = timestampMs ?: nowMs()
+        if (!endWithFrameDraw) {
+            dataCollectionComplete()
+        }
+    }
+
+    fun firstFrameRendered(timestampMs: Long? = null) {
+        firstFrameRenderedMs = timestampMs ?: nowMs()
+        if (endWithFrameDraw) {
+            dataCollectionComplete()
+        }
+    }
+
+    private fun dataCollectionComplete() {
+        if (!startupRecorded.get()) {
+            synchronized(startupRecorded) {
+                if (!startupRecorded.get()) {
+                    backgroundWorker.submit {
```

Comment:
Not sure about recording this on a background thread as it's hard to tell when it will get scheduled, particularly if the device is under stress

Reply:
Made it not use a background thread for now

Reply:
lol I waffled. I AM going to use a background thread here to minimize impact to the UI thread. Since this is not in production, I'm ok with losing some data potentially, but I don't want to cause additional processing on the UI thread if I don't have to